src/TemporaryFile.cpp

```diff
@@ -42,10 +43,23 @@ get_umask()
 } // namespace
 
 TemporaryFile::TemporaryFile(string_view path_prefix)
-  : path(std::string(path_prefix) + ".XXXXXX")
 {
-  Util::ensure_dir_exists(Util::dir_name(path));
+  Util::ensure_dir_exists(Util::dir_name(path_prefix));
+
+#ifdef _WIN32
+  LARGE_INTEGER timestamp{};
+  QueryPerformanceCounter(&timestamp);
```

Comment:
Can we be sure that the timestamp never is the same for two calls to `TemporaryFile` quickly after each other?

Reply:
In theory, this is indeed a problem. In practice, anything that touches the filesystem on Windows is going to take a few dozen microseconds at least, so you'll get a different timestamp from each `QueryPerformanceCounter(); open()/CreateFile()` call.

That said, my preference would still be along the lines of the original PR, i.e. generate a random file name using a good PRNG, or generate a UUID and use that.

Reply:
> In theory, this is indeed a problem. [...] That said, my preference would still be along the lines of the original PR, i.e. generate a random file name using a good PRNG, or generate a UUID and use that.

That sounds like hoping, not guaranteeing. I would prefer an implementation that doesn't even have the theoretical problem. How about some of these alternatives?

Alternative A: Add PID to the filename and use `mkstemp()`. That should be enough since if `mkstemp()` returns a recently deleted file, it won't be open due to the PID in the name (the previous process with our PID no longer exists). Bonus: To improve probing efficiency, do something like `srand(timestamp)` at program start.

Alternative B: Your current proposal but retry until `fd >= 0 || (fd != EEXIST && fd != EACCES)` plus some `RtlGetLastNtStatus()` check or similar to distinguish between the `EACCES` cases.

I would prefer alternative A unless you see something wrong with it.

Reply:
On the other hand, I guess alternative B would be preferred to keep people who want to build ccache with MSVC happy?

Reply:
I can code up both and see which you prefer. I understand your reluctance to merge a bunch of Windows-only code given that ccache has historically only supported POSIX platforms. If ccache is intended to be single-threaded forevermore and only support Windows builds via mingw, then calling `srand()` in `main()` should fix the issue and not introduce much Windows-specific code.

But how about a `int Util::unique_fd(char*)` that calls a custom implementation on Windows but `int mkstemp(char*)` on other platforms? IMHO supporting Windows means avoiding a dependency on mingw -- I'd be inclined to distrust mingw-w64 and its portability wrappers after seeing how poorly they implemented `mkstemp()` ;) It looks like ccache relying on `mkstemp` is actually only thing that is preventing it being built under MSVC. GitHub actions supports building with MSVC (or Clang using the MSVC stdlib). Happy to contribute the CI glue to add support for this.

Reply:
> But how about a `int Util::unique_fd(char*)` that calls a custom implementation on Windows but `int mkstemp(char*)` on other platforms?

Sure.

> IMHO supporting Windows means avoiding a dependency on mingw -- I'd be inclined to distrust mingw-w64 and its portability wrappers after seeing how poorly they implemented mkstemp() ;)

I have no opinion on this myself, other than that I would like to continue being able to cross-compile with Mingw-w64 and run unit tests in Wine locally since I don't have a Windows machine myself.

> It looks like ccache relying on `mkstemp` is actually only thing that is preventing it being built under MSVC.

That's because there was a drive in #632 to make it build with MSVC (this is much newer than the Windows support via Mingw-w64, and via MSYS2 even earlier) and I removed the fallback to `mktemp()` recently in 899a1025c287c57df25ac638d159fc03f075fd3f.

> GitHub actions supports building with MSVC (or Clang using the MSVC stdlib). Happy to contribute the CI glue to add support for this.

Sure, that would be likely be valuable. It was mentioned in #632 as well but nobody followed up on it.